src/LanguageServer/Impl/CodeActions/MissingImportCodeActionProvider.cs

```diff
@@ -0,0 +1,635 @@
+using Microsoft.Python.Analysis.Values;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.Collections;
+using Microsoft.Python.Core.Text;
+using Microsoft.Python.LanguageServer.Diagnostics;
+using Microsoft.Python.LanguageServer.Indexing;
+using Microsoft.Python.LanguageServer.Protocol;
+using Microsoft.Python.Parsing.Ast;
+using Range = Microsoft.Python.Core.Text.Range;
+
+namespace Microsoft.Python.LanguageServer.CodeActions {
+    internal sealed class MissingImportCodeActionProvider : ICodeActionProvider {
+        private const int ModuleLoadTimeout = 10 * 1000; // 10 seconds
+
+        public static readonly ICodeActionProvider Instance = new MissingImportCodeActionProvider();
+
+        // right now, it is a static. in future, we might consider giving an option to users to customize this list
+        // also, right now, it is text based. so if module has same name, they will get same suggestion even if
+        // the module is not something user expected
+        private static readonly Dictionary<string, string> WellKnownAbbreviationMap = new Dictionary<string, string>() {
+            { "numpy", "np" },
+            { "pandas", "pd" }
+        };
+
+        private MissingImportCodeActionProvider() {
+        }
+
+        public ImmutableArray<string> FixableDiagnostics => ImmutableArray<string>.Create(
+            ErrorCodes.UndefinedVariable, ErrorCodes.VariableNotDefinedGlobally, ErrorCodes.VariableNotDefinedNonLocal);
+
+        public async Task<IEnumerable<CodeAction>> GetCodeActionAsync(IDocumentAnalysis analysis, DiagnosticsEntry diagnostic, CancellationToken cancellationToken) {
+            var finder = new ExpressionFinder(analysis.Ast, FindExpressionOptions.Complete);
+            var node = finder.GetExpression(diagnostic.SourceSpan);
+            if (!(node is NameExpression)) {
+                return Enumerable.Empty<CodeAction>();
+            }
+
+            var interpreter = analysis.Document.Interpreter;
+            var pathResolver = interpreter.ModuleResolution.CurrentPathResolver;
+
+            var name = node.ToCodeString(analysis.Ast);
+            if (string.IsNullOrEmpty(name)) {
+                return Enumerable.Empty<CodeAction>();
+            }
+
+            await EnsureCandidateModulesLoadedAsync(analysis, name, cancellationToken);
+
+            var importFullNameMap = new Dictionary<string, ImportInfo>();
+
+            // find installed modules matching the given name. this will include submodules
+            var languageVersion = Parsing.PythonLanguageVersionExtensions.ToVersion(interpreter.LanguageVersion);
+            var includeImplicit = !ModulePath.PythonVersionRequiresInitPyFiles(languageVersion);
+            foreach (var moduleFullName in pathResolver.GetAllImportableModulesByName(name, includeImplicit)) {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var module = interpreter.ModuleResolution.GetOrLoadModule(moduleFullName);
```

Comment:
This is extremely heavy, loading module causes analysis and will load the universe. Should only be used if module is actually imported by the user directly or indirectly.

Reply:
for things like os.path, I dont think we can support it without analyzing the module. we could have cache created for libraries but right now it doesn't exist. so that is not an option for now.

so, how much this different than find all references? find all references doesn't load libraries?

Reply:
Find references does load modules and I think it actually does too much - which may explain some bugs about memory consumption. IMO, after finding root definition it should only be scanning user code. I don't think user is interested in references elsewhere.

Reply:
I showed him what happens in my awful conda environment from the PyQt bug. We're going to try and drop this for now and rely only on indexing.

Reply:
I meant find all references but nvm, it is limited to root which should be just user code.

Reply:
will make the change and send out update